### DISCUSSION TITLE
Please help! UDM / UDMPro Boot Script

### DISCUSSION THREAD

**Discussion by Nick-MAC** [unifi-utilities/unifi-common#266] [Q&A]
I consider myself a pretty smart guy and i have been researching this for some time with no success so hoping someone will be bale to point me in the right direction. First, I'm starting to get a little frustrated with all of the things this "PRO" router does not come with OOTB but that aside, i am bought in at this point.

I followed all of the directions for the setup of this Boot Utilities, but no matter what i do my script does not start on boot. I am able to start my script manually and it works just fine, i have the correct #!/bin/sh , I have the file in the correct directory, i have made my script executable. My script is similar to the following. My ultimate goal is to auto connect one of my VLANs to a VPN. Any help would be fantastic. 

I don't think my issue is with the script below, i was able to follow the install instructions for the boot service fairly easily. Honestly no idea where i need to look to find the issue.

#!/bin/sh
cd /mnt/data/split-vpn/openvpn/pia
source ./vpn.conf
/mnt/data/split-vpn/vpn/updown.sh ${DEV} pre-up &> pre-up.log
nohup openvpn --config filename.ovpn \
              --route-noexec \
              --up /mnt/data/split-vpn/vpn/updown.sh \
              --down /mnt/data/split-vpn/vpn/updown.sh \
              --auth-user-pass username_password.txt \
              --dev-type tun --dev ${DEV} \
              --script-security 2 \
              --ping-restart 15 \
              --mute-replay-warnings &> openvpn.log &

**Comment by boostchicken** [comment]
Well the first thing you can do to see if the startup hook is running is get into unifios shell and look at the status of the udm-boot.service. If that is running it should be fine.  Does this file end .sh?  It should otherwise it will get filtered

**Reply by Nick-MAC** [reply]
Thanks for checking in! Yes, my file does indeed end in .sh and I placed it in here "/mnt/data/on_boot.d"

I am new to the Unifi Shell stuff so could you point me in the right direction for ensuring the service has started that would be much obliged.

**Reply by Nick-MAC** [reply]
ok i figured out the right command. that's where the problem is. there is no udm-boot.service. Should that not install automagically when i run the following?

unifi-os shell
curl -L https://udm-boot.boostchicken.dev -o udm-boot_1.0.5_all.deb
dpkg -i udm-boot_1.0.5_all.deb

**Reply by Nick-MAC** [reply]
ok more info... I removed the package and the added and now i have this. Any idea on what i should do here?

● udm-boot.service                                                                 loaded failed     failed    Run On Startup UDM

**Reply by Nick-MAC** [reply]
ok got some more info for you after reading through some of the other posts. I am on controller version 1.10.4. Below is the output of journalctl -u udm-boot.service. Hoping this will help you to point me in the right direction


root@ubnt:/# journalctl -u udm-boot.service
-- Logs begin at Tue 2021-11-16 22:39:02 CST, end at Tue 2021-11-16 22:46:52 CST. --
Nov 16 22:44:58 ubnt systemd[1]: Starting Run On Startup UDM...
Nov 16 22:44:58 ubnt ssh-proxy[9119]: udm-boot.service: running /mnt/data/on_boot.d/99-run-vpn.sh
Nov 16 22:44:58 ubnt ssh-proxy[9119]: /mnt/data/on_boot.d/99-run-vpn.sh: line 1: /mnt/data/on_boot.d/99-run-vpn.sh: not found
Nov 16 22:44:58 ubnt systemd[1]: udm-boot.service: Main process exited, code=exited, status=123/n/a
Nov 16 22:44:58 ubnt systemd[1]: udm-boot.service: Failed with result 'exit-code'.
Nov 16 22:44:58 ubnt systemd[1]: Failed to start Run On Startup UDM.
root@ubnt:/#

**Reply by Nick-MAC** [reply]
Any chance you have some insight as to what is going on here? Is it possible there is a problem with v1.05 and my current controller version? Outside of that I'm not sure what this could be, as i followed the steps very closely and do not think i did anything incorrectly during the setup process.

**Reply by VeniceNerd** [reply]
> Well the first thing you can do to see if the startup hook is running is get into unifios shell and look at the status of the udm-boot.service. If that is running it should be fine. Does this file end .sh? It should otherwise it will get filtered

What command do I need to run to check this? I'm also having issues and trying to trouble shoot.

**Reply by VeniceNerd** [reply]
> ok i figured out the right command. that's where the problem is. there is no udm-boot.service. Should that not install automagically when i run the following?
> 
> unifi-os shell curl -L https://udm-boot.boostchicken.dev -o udm-boot_1.0.5_all.deb dpkg -i udm-boot_1.0.5_all.deb

Would you be so kind to let me know the command you ran? I'm having issued myself and would love to see if everything is even running. Thank you!

**Reply by Nick-MAC** [reply]
Would be nice if I could get some help on this, as this is still not working for me... @boostchicken ? Any ideas what may be going on here? I followed all of the steps exactly and rinsed repeated and still no joy... The service never starts.

That aside, below are the commands you will need to run to check the service.

1. You need to SSH to the UDM Pro
2. run - "unifi-os shell"
3. run - "systemctl"
4. Then you can look for the "udm-boot.service" and see what its status it.

![image](https://user-images.githubusercontent.com/54548626/144458995-6e2afc80-af58-4df6-a595-c6b78a1064b4.png)

**Reply by VeniceNerd** [reply]
> 1. You need to SSH to the UDM Pro
> 2. run - "unifi-os shell"
> 3. run - "systemctl"
> 4. Then you can look for the "udm-boot.service" and see what its status it.

Thank you so much! Mine shows the following:

<img width="837" alt="Screenshot 2021-12-02 at 09 47 56" src="https://user-images.githubusercontent.com/25128243/144476089-f1f33b36-6aa6-42bd-a3fa-1b1e876556cb.png">

Would that be good news, even though it says "exited"?

I really hope someone will respond to your issue soon. The community seems really small so it's very hard to come by any answers.

**Reply by Nick-MAC** [reply]
LOL, yes... Very small community indeed. Honestly, I'm at the point where I will try a factory reset. Kind of frustrating, but may be all I can do at this point. I've been a Systems Admin for 15 years now so I'm not an idiot, but I'm no linux/networking wiz, so i'm pretty much at the extent of my knowledge on  this stuff   ¯\_(ツ)_/¯

You are in a better position than me given that your service is in a good state. I would think that you would want it to be started. You can use the same "systemctl" to start said service

1. run - "systemctl start udm-boot.service"
2.  then check to see if the service is a running state. using "systemctl"
3.  Then try a reboot and your.sh file should kick in and your VPN automatically connected.

Side note - I've mentioned this before, but I cant believe that this is not an OOTB service! This is a "Pro" router after all, lol.

**Reply by VeniceNerd** [reply]
No dice. Ran the command:
<img width="437" alt="Screenshot 2021-12-02 at 13 53 31" src="https://user-images.githubusercontent.com/25128243/144509226-4878e4ae-ccf3-465f-9d70-c2fe2edbb6eb.png">

Same result:

<img width="575" alt="Screenshot 2021-12-02 at 13 54 21" src="https://user-images.githubusercontent.com/25128243/144509299-68c2c5a5-cd7c-4215-a1a0-7d74179ee198.png">


Are we sure, though, that it needs to show as "loaded / active/ running " or is it maybe normal that it says "loaded / active / exited"? Maybe @boostchicken will have some time to chime in at some point. Would be great to know.

**Comment by drmugzy** [comment]
### Troubleshooting udm-boot.service in [systemd](https://systemd.io/)
@nickmac88, I know this is a bit late, but I came upon your post while troubleshooting this same issue on my UDM-Pro.  @VeniceNerd was on the right track, troubleshooting the systemd unit file. While both of you have probably already resolved the issue, I wanted to post in case it helps others and to offer an answer, closing this thread (hopefully).

Without going on too much about systemd, I'd like to highlight a few commands that would've helped ya'll zero in on the culprit. In this specific scenario, after @boostchicken pointed to the "udm-boot.service" unit file, my troubleshooting process proceeded as follows:

1. Enter the UniFi-OS shell.

     `unifi-os shell`

2. Check the status of the _udm-boot.service_ unit file.

     `systemctl status udm-boot` (The ".service" file extension is assumed by systemd)

     In my case, this service was "disabled", which is why my boot scripts weren't running at startup.

     
![image](https://user-images.githubusercontent.com/44238679/156864864-403c2f12-7c86-4581-9607-0a55081ec266.png)

3. Ensure the _udm-boot.service_ unit file is "enabled" (in systemd "enabled" means that it runs at startup).

     `systemctl enable udm-boot`

     After verifying the unit file was actually enabled, all startup scripts executed as expected. To verify:

     `systemctl status udm-boot`

     
![image](https://user-images.githubusercontent.com/44238679/156865276-eed17041-2981-46dc-830d-3806537ccc31.png)

4. To continue troubleshooting, examine the _udm-boot.service_ unit file.

     `systemctl cat udm-boot`

     
![image](https://user-images.githubusercontent.com/44238679/156865538-5a2f84a2-25fc-49b5-9630-660d3b09e9e0.png)

   Here we start to understand how the service functions, within the UniFi-OS container.  @VeniceNerd , to answer your last question, "loaded, active, exited" is the appropriate status. The type "oneshot" indicates that the service will run once and exit. Digital Ocean has a pretty good [systemd overview](https://www.digitalocean.com/community/tutorials/understanding-systemd-units-and-unit-files) with practical explanations.

5.  If the desired behavior isn't achieved after rebooting, check the systemd journal for errors.

     `journalctl -xeu udm-boot`  (x=expanded output, useful messages when available | e=end of log | u=unit to look for)

      @nickmac88, your journal indicated an issue with 99-run-vpn.sh on line 1, specifically that it's missing.

> Nov 16 22:44:58 ubnt ssh-proxy[9119]: /mnt/data/on_boot.d/99-run-vpn.sh: line 1: /mnt/data/on_boot.d/99-run-vpn.sh: not found

@nickmac88, your issue seems be an invalid directory structure. The _udm-boot.service_ will only traverse the target directory to a depth of 1, as observed in the "ExecStart="  directive of the unit file.  If you ran the setup script  within the UniFi-OS container, the appropriate directory structure would be missing. The persistence of /mnt/data/ is only achieved on the linux host, not inside of any of the application containers, thus the _udm-boot.service_ pulls the startup scripts from the linux host and into the UniFi-OS application container.

@nickmac88 , your installation command indicates you ran the setup from within the UniFi-OS application container.

> unifi-os shell
 curl -L https://udm-boot.boostchicken.dev/ -o udm-boot_1.0.5_all.deb
dpkg -i udm-boot_1.0.5_all.deb

@nickmac88 , pay close attention to the terminal prompt you're working in.

     " # " indicates that your operating on the linux host

     " root@ubnt:/# "  indicates that your operating within the UniFi-OS application container
